falcon/falcon.py

```diff
@@ -246,231 +293,266 @@ def main(args: Union[str, List[str]] = None) -> int:
-            spectra_queue.put(spec)
-
+    charge : int
+        The precursor charge of the spectra.
+    schema : pa.Schema
+        The schema of the dataset.
 
-def _read_spectra(filename: str, mz_interval: int, work_dir: str) \
-        -> Iterator[Tuple[MsmsSpectrum, str]]:
+    Returns
+    -------
+    lance.LanceDataset
+        The lance dataset.
+    """
+    lance_path = os.path.join(
+        config.work_dir, "spectra", f"spectra_charge_{charge}.lance"
+    )
+    dataset = lance.write_dataset(
+        pa.Table.from_pylist([], schema),
+        lance_path,
+        mode="overwrite",
+        data_storage_version="stable",
+    )
+    logger.debug("Creating lance dataset at %s", lance_path)
+    return dataset
+
+
+def _read_spectra(
+    filename: str,
+    process_spectrum: Callable,
+    low_quality_counter: multiprocessing.sharedctypes.Synchronized,
+) -> List[Dict[str, Union[str, float, int, np.ndarray]]]:
     """
     Get the spectra from the given file.
 
     Parameters
     ----------
     filename : str
         The path of the peak file to be read.
-    mz_interval : int
-        The width of each m/z interval.
-    work_dir : str
-        The directory in which the spectrum buckets will be stored.
+    process_spectrum : Callable
+        The function to process the spectra.
+    low_quality_counter : multiprocessing.sharedctypes.Synchronized
+        Counter for low-quality spectra.
 
     Returns
     -------
-    Iterator[Tuple[MsmsSpectrum, str]]
-        The spectra read from the given file and their bucket filenames
-        (based on precursor charge and m/z).
+    List[Dict[str, Union[str, float, int, np.ndarray]]]
+        The spectra read from the given file as a list of dictionaries.
     """
+    spectra = []
     filename = os.path.abspath(filename)
     for spec in ms_io.get_spectra(filename):
         spec.filename = filename
-        interval = _precursor_to_interval(
-            spec.precursor_mz, spec.precursor_charge, mz_interval)
-        pkl_filename = os.path.join(
-            work_dir, 'spectra', f'{spec.precursor_charge}_{interval}.pkl')
-        yield spec, pkl_filename
-
-
-@nb.njit(cache=True)
-def _precursor_to_interval(mz: float, charge: int, interval_width: int) -> int:
-    """
-    Convert the precursor m/z to the neutral mass and get the interval index.
-
-    Parameters
-    ----------
-    mz : float
-        The precursor m/z.
-    charge : int
-        The precursor charge.
-    interval_width : int
-        The width of each m/z interval.
-
-    Returns
-    -------
-    int
-        The index of the interval to which a spectrum with the given m/z and
-        charge belongs.
-    """
-    hydrogen_mass, cluster_width = 1.00794, 1.0005079
-    neutral_mass = (mz - hydrogen_mass) * max(abs(charge), 1)
-    return round(neutral_mass / cluster_width) // interval_width
+        spec = process_spectrum(spec)
+        if spec is None:
+            low_quality_counter.value += 1
```

Comment:
**question:** Is there any benefit if first accumulating the number of skipped spectra in a local variable and then only updating the shared `low_quality_counter`, which needs to be synchronized across threads, once at the end?

Reply:
If I remember correctly, constantly updating the shared variable made it slower

Reply:
ah wait, that's what I'm doing 😅 Thought I already changed this